### PR TITLE
No panics in config constructors

### PR DESCRIPTION
**Pull request by alsamylkin**
Related to https://github.com/uber-go/config/issues/24

Instead of panicking on errors, we should simply return them
to a caller end let them decide what to do with the errors.

The change is mostly straight forward, but there is one snowflake:
expand variables. The replace function doesn't return an error,
so we panic internally and use recover to report errors.

### CODE REVIEW

File: yaml.go [hunk 1]
```diff
@@ -285,18 +320,29 @@ func (n *yamlNode) Children() []*yamlNode {
...
 
-func (n *yamlNode) applyOnAllNodes(expand func(string) string) {
-	if n == nil {
-		return
-	}
+func (n *yamlNode) applyOnAllNodes(expand func(string) string) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			if _, ok := r.(runtime.Error); ok {
+				fmt.Println(r)
```
Comment (glibsm) [inline yaml.go]: no need to print before panic here

File: command_line_provider.go [hunk 2]
```diff
@@ -77,7 +77,12 @@ func NewCommandLineProvider(flags *flag.FlagSet, args []string) Provider {
 		m[f.Name] = f.Value.String()
 	})
 
-	return commandLineProvider{Provider: NewStaticProvider(m)}
+	p, err := NewStaticProvider(m)
+	if err != nil {
+		return nil, err
```
Comment (glibsm) [inline command_line_provider.go]: not covered

File: static_provider.go [hunk 3]
```diff
@@ -35,33 +35,48 @@ type staticProvider struct {
...
-		Provider: NewYAMLProviderFromReader(toReadCloser(data)),
+func NewStaticProvider(data interface{}) (Provider, error) {
+	c, err := toReadCloser(data)
+	if err != nil {
+		return nil, err
 	}
+
+	p, err := NewYAMLProviderFromReader(c)
+	if err != nil {
+		return nil, err
```
Comment (glibsm) [inline static_provider.go]: also not covered

File: yaml.go [hunk 4]
```diff
@@ -285,18 +320,29 @@ func (n *yamlNode) Children() []*yamlNode {
...
 }
 
-func (n *yamlNode) applyOnAllNodes(expand func(string) string) {
-	if n == nil {
-		return
-	}
+func (n *yamlNode) applyOnAllNodes(expand func(string) string) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			if _, ok := r.(runtime.Error); ok {
```
Comment (glibsm) [inline yaml.go]: do we want to explicitly filter out the runtime errors in this case?

Comment (alsamylkin) [inline yaml.go]: yes, runtime.Error is returned for nil pointer exceptions/index out of bounds. Config internal errors are plain builtin errors.

File: yaml.go (hunk 1 above)
Comment (alsamylkin) [inline yaml.go]: oops..

File: command_line_provider.go (hunk 2 above)
Comment (alsamylkin) [inline command_line_provider.go]: I can cover this line, but it'll not provide much value, simply a public function returning an error via another public function.

File: static_provider.go (hunk 3 above)
Comment (alsamylkin) [inline static_provider.go]: This is almost impossible to cover and doesn't provide much value.

**Comment by alsamylkin** [comment]
image

**Pull request merged**